discovery/dockerswarm/services.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dockerswarm
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strconv"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/swarm"
+	"github.com/prometheus/common/model"
+
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+	"github.com/prometheus/prometheus/util/strutil"
+)
+
+const (
+	swarmLabelServicePrefix                  = swarmLabel + "service_"
+	swarmLabelServiceEndpointPortName        = swarmLabelServicePrefix + "endpoint_port_name"
+	swarmLabelServiceEndpointPortPublishMode = swarmLabelServicePrefix + "endpoint_port_publish_mode"
+	swarmLabelServiceID                      = swarmLabelServicePrefix + "id"
+	swarmLabelServiceLabelPrefix             = swarmLabelServicePrefix + "label_"
+	swarmLabelServiceName                    = swarmLabelServicePrefix + "name"
+	swarmLabelServiceMode                    = swarmLabelServicePrefix + "mode"
+	swarmLabelServiceUpdatingStatus          = swarmLabelServicePrefix + "updating_status"
+	swarmLabelServiceTaskPrefix              = swarmLabelServicePrefix + "task_"
+	swarmLabelServiceTaskContainerImage      = swarmLabelServiceTaskPrefix + "container_image"
+	swarmLabelServiceTaskContainerHostname   = swarmLabelServiceTaskPrefix + "container_hostname"
+
+	swarmLabelValueModeGlobal     = model.LabelValue("global")
+	swarmLabelValueModeReplicated = model.LabelValue("replicated")
+)
+
+func (d *Discovery) refreshServices(ctx context.Context) ([]*targetgroup.Group, error) {
+	tg := &targetgroup.Group{
+		Source: "DockerSwarm",
+	}
+
+	services, err := d.client.ServiceList(ctx, types.ServiceListOptions{Filters: makeArgs(d.filters)})
+	if err != nil {
+		return nil, fmt.Errorf("error while listing swarm services: %w", err)
+	}
+
+	networkLabels := make(map[string]model.LabelSet, 1)
+
+	for _, s := range services {
+		for _, e := range s.Endpoint.Ports {
+			if e.Protocol != swarm.PortConfigProtocolTCP {
+				continue
+			}
+			for _, p := range s.Endpoint.VirtualIPs {
+				labels := model.LabelSet{
+					swarmLabelServiceEndpointPortName:        model.LabelValue(e.Name),
+					swarmLabelServiceEndpointPortPublishMode: model.LabelValue(e.PublishMode),
+					swarmLabelServiceID:                      model.LabelValue(s.ID),
+					swarmLabelServiceMode:                    getServiceValueMode(s),
```

Comment:
Generally we'd inline this, rather than hiding it among the simple string assignments.

Reply:
inlined.